### PR TITLE
DAOS-6358 Test: Pool destroy failed with DER_TIMEOUT

### PR DESCRIPTION
**Pull request by omar-ocampo**
Added function to wait for pool rebuild to complete.
    Pool is now destroyed within the test.

Skip-unit-tests: true
Skip-coverity-test: true
Test-tag-vm: pr,-hw,test_pool_svc,-hw
Signed-off-by: Omar Ocampo <omar.ocampo.coronado@intel.com>

### CODE REVIEW

**Comment by omar-ocampo** [comment]
Test passing: https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-5358/3/tests

File: src/tests/ftest/pool/pool_svc.py [hunk 1]
```diff
@@ -130,6 +166,8 @@ def test_pool_svc(self):
                         "DaosServerManager.stop_ranks([{}])".format(
                             pool_leader))
 
+                self.is_pool_rebuild_in_progress()
```
Comment (cibervicho) [inline src/tests/ftest/pool/pool_svc.py:169]: This function returns a bool, but seems you are not taking that into account here. How do you determine that something goes wrong if the number of retries (defaulted to 5) is reached and never changed the state to done?

Comment (omar-ocampo) [reply src/tests/ftest/pool/pool_svc.py:169]: In case the rebuild is still in progress, deleting the pool will fail.

Comment (cibervicho) [reply src/tests/ftest/pool/pool_svc.py:169]: Got it

**Review by cibervicho — APPROVED** [review]

**Comment by omar-ocampo** [comment]
Bring this to the rebuild WG

**Comment by omar-ocampo** [comment]
There was one failure not related to this PR: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5358/5/artifact/Functional%20on%20CentOS%207/pool/pool_svc.py/results.json 

There was no issue when destroying pool

**Comment by omar-ocampo** [comment]
Leap: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5358/5/artifact/Functional%20on%20Leap%2015/pool/pool_svc.py/results.json

Centos: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5358/5/artifact/Functional%20on%20CentOS%207/pool/pool_svc.py/results.json

File: src/tests/ftest/pool/pool_svc.py [hunk 2]
```diff
@@ -55,6 +56,42 @@ def check_leader(self, previous_leader=None, expect_change=True):
             self.log.info("Pool leader: current=%s", current_leader)
         return current_leader
 
+    def is_pool_rebuild_in_progress(self, max_retries=5, sleep_timer=5):
```
Comment (shimizukko) [inline src/tests/ftest/pool/pool_svc.py:59]: Not sure what your intent here. Are you trying to wait until rebuild is done? If so, you can use `TestPool.wait_for_rebuild()`. or are you trying to do something while the state is "rebuild in progress"? If so, why are you not using the return value?

Comment (omar-ocampo) [reply src/tests/ftest/pool/pool_svc.py:59]: I wasnt aware of the function wait_for_rebuild, the functions are quite similar. Changing code to use the function already coded.

**Review by shimizukko — DISMISSED** [review]

**Comment by sylviachanoiyee** [comment]
@omar-ocampo , the issue described in DAOS-6358 is on test pool_list failing.
Why  is the fix in this PR on pool svc?

**Comment by omar-ocampo** [comment]
> @omar-ocampo , the issue described in DAOS-6358 is on test pool_list failing.
> Why is the fix in this PR on pool svc?

The initial report shows the failure in test_pool_list, just as you say. As time passed the issue was fixed in pool_list and it appeared also in pool svc, see this comment: https://jira.hpdd.intel.com/browse/DAOS-6358?focusedCommentId=259726&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-259726

**Comment by omar-ocampo** [comment]
@sylviachanoiyee  friendly reminder

**Comment by sylviachanoiyee** [comment]
> > @omar-ocampo , the issue described in DAOS-6358 is on test pool_list failing.
> > Why is the fix in this PR on pool svc?
> 
> The initial report shows the failure in test_pool_list, just as you say. As time passed the issue was fixed in pool_list and it appeared also in pool svc, see this comment: https://jira.hpdd.intel.com/browse/DAOS-6358?focusedCommentId=259726&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-259726

Thanks. Need another reviewer

**Comment by sylviachanoiyee** [comment]
@saurabhtandan , @rpadma2 , could you help review this PR since you guys are familiar with changes in pool test with rebuild? Thanks!

**Review by saurabhtandan — COMMENTED** [review]
@omar-ocampo , any specific reason to add pool destroy as part of test run and not leave it for teardown?
Also, I noticed the original issue in the ticket, created a list of pools and triggered rebuild and then on pool destroy there was a timeout.
I understand with time the test might have evolved and there is no scenario in the test variants now where multiple pools are created as part of same test variant and then rebuild is triggered.
Do you think we should add that scenario and then test for pool destroy timeout?

File: src/tests/ftest/pool/pool_svc.py [hunk 3]
```diff
@@ -130,6 +129,8 @@ def test_pool_svc(self):
                         "DaosServerManager.stop_ranks([{}])".format(
                             pool_leader))
 
+                self.pool.wait_for_rebuild(False, interval=1)
```
Comment (rpadma2) [inline src/tests/ftest/pool/pool_svc.py:132]: Minor: I believe you don't need to pass interval=1. The default is 1.

File: src/tests/ftest/pool/pool_svc.py [hunk 4]
```diff
@@ -148,7 +149,9 @@ def test_pool_svc(self):
                         "Error stopping a pool non-leader - "
                         "DaosServerManager.stop_ranks([{}])".format(non_leader))
 
+                self.pool.wait_for_rebuild(False, interval=1)
```
Comment (rpadma2) [inline src/tests/ftest/pool/pool_svc.py:152]: Same as above.

File: src/tests/ftest/pool/pool_svc.py [hunk 5]
```diff
@@ -148,7 +149,9 @@ def test_pool_svc(self):
...
                 # Verify the pool leader has not changed
                 self.check_leader(pool_leader, False)
+                self.destroy_pools(self.pool)
```
Comment (rpadma2) [inline src/tests/ftest/pool/pool_svc.py:155]: Agree with Saurabh. Any reason why we destroy the pools here. Teardown should handle it.

**Comment by omar-ocampo** [comment]
> @omar-ocampo , any specific reason to add pool destroy as part of test run and not leave it for teardown?
> Also, I noticed the original issue in the ticket, created a list of pools and triggered rebuild and then on pool destroy there was a timeout.
> I understand with time the test might have evolved and there is no scenario in the test variants now where multiple pools are created as part of same test variant and then rebuild is triggered. (if I missed and there is one can you please point me to it)
> Do you think we should add that scenario and then test for pool destroy timeout?

Good question. The idea of moving the pool destroy within the test instead of the teardown is: if in the future we have the same/similar failure we can pinpoint the issue within the test and not the teardown. While the difference is subtle it may make a big difference in debugging. You are correct, if we speak in terms of functionality it would be the same (delete within test or teardown) I just favor the maintainability of the code. 

Having say that, I'm open of letting the teardown to delete the pool, should we keep the pool delete within the test or moved it out?

Yes, only one pool is created. The rebuild testing is being done within pool/rebuild_tests.py, should be easier to add "--svn" as a test param and be able to use the logic of write data, stop server, rebuild pool and read data.

File: src/tests/ftest/pool/pool_svc.py (hunk 5 above)
Comment (omar-ocampo) [reply src/tests/ftest/pool/pool_svc.py:155]: Just answered to Saurabh in this comment: https://github.com/daos-stack/daos/pull/5358#issuecomment-820772434 (it should be just above this one).

File: src/tests/ftest/pool/pool_svc.py (hunk 3 above)
Comment (omar-ocampo) [reply src/tests/ftest/pool/pool_svc.py:132]: Actually I'd like to change it to:
    self.pool.wait_for_rebuild(to_start=False, interval=1) for the sake of readability.

**Comment by omar-ocampo** [comment]
> > @omar-ocampo , any specific reason to add pool destroy as part of test run and not leave it for teardown?
> > Also, I noticed the original issue in the ticket, created a list of pools and triggered rebuild and then on pool destroy there was a timeout.
> > I understand with time the test might have evolved and there is no scenario in the test variants now where multiple pools are created as part of same test variant and then rebuild is triggered. (if I missed and there is one can you please point me to it)
> > Do you think we should add that scenario and then test for pool destroy timeout?
> 
> Good question. The idea of moving the pool destroy within the test instead of the teardown is: if in the future we have the same/similar failure we can pinpoint the issue within the test and not the teardown. While the difference is subtle it may make a big difference in debugging. You are correct, if we speak in terms of functionality it would be the same (delete within test or teardown) I just favor the maintainability of the code.
> 
> Having say that, I'm open of letting the teardown to delete the pool, should we keep the pool delete within the test or moved it out?
> 
> Yes, only one pool is created. The rebuild testing is being done within pool/rebuild_tests.py, should be easier to add "--svn" as a test param and be able to use the logic of write data, stop server, rebuild pool and read data.

@rpadma2 @saurabhtandan, thoughts on this?

**Comment by sylviachanoiyee** [comment]
The fn vm test stage failed

**Comment by omar-ocampo** [comment]
@rpadma2 @saurabhtandan @sylviachanoiyee ,  friendly reminder, anything I can help with?

**Comment by omar-ocampo** [comment]
@rpadma2 @saurabhtandan @sylviachanoiyee , anything I can help with?

**Comment by rpadma2** [comment]
> > > @omar-ocampo , any specific reason to add pool destroy as part of test run and not leave it for teardown?
> > > Also, I noticed the original issue in the ticket, created a list of pools and triggered rebuild and then on pool destroy there was a timeout.
> > > I understand with time the test might have evolved and there is no scenario in the test variants now where multiple pools are created as part of same test variant and then rebuild is triggered. (if I missed and there is one can you please point me to it)
> > > Do you think we should add that scenario and then test for pool destroy timeout?
> > 
> > 
> > Good question. The idea of moving the pool destroy within the test instead of the teardown is: if in the future we have the same/similar failure we can pinpoint the issue within the test and not the teardown. While the difference is subtle it may make a big difference in debugging. You are correct, if we speak in terms of functionality it would be the same (delete within test or teardown) I just favor the maintainability of the code.
> > Having say that, I'm open of letting the teardown to delete the pool, should we keep the pool delete within the test or moved it out?
> > Yes, only one pool is created. The rebuild testing is being done within pool/rebuild_tests.py, should be easier to add "--svn" as a test param and be able to use the logic of write data, stop server, rebuild pool and read data.
> 
> @rpadma2 @saurabhtandan, thoughts on this?

How often this problem happens under CI? Is it only in this test alone. I am not sure whether adding pool.destroy within the test should resolve the issue. If @saurabhtandan is ok with the change, we can land it... I will also check with @phender (Phil) to understand why it failed based on the existing logs. He can provide some input...

**Comment by omar-ocampo** [comment]
> > > > @omar-ocampo , any specific reason to add pool destroy as part of test run and not leave it for teardown?
> > > > Also, I noticed the original issue in the ticket, created a list of pools and triggered rebuild and then on pool destroy there was a timeout.
> > > > I understand with time the test might have evolved and there is no scenario in the test variants now where multiple pools are created as part of same test variant and then rebuild is triggered. (if I missed and there is one can you please point me to it)
> > > > Do you think we should add that scenario and then test for pool destroy timeout?
> > > 
> > > 
> > > Good question. The idea of moving the pool destroy within the test instead of the teardown is: if in the future we have the same/similar failure we can pinpoint the issue within the test and not the teardown. While the difference is subtle it may make a big difference in debugging. You are correct, if we speak in terms of functionality it would be the same (delete within test or teardown) I just favor the maintainability of the code.
> > > Having say that, I'm open of letting the teardown to delete the pool, should we keep the pool delete within the test or moved it out?
> > > Yes, only one pool is created. The rebuild testing is being done within pool/rebuild_tests.py, should be easier to add "--svn" as a test param and be able to use the logic of write data, stop server, rebuild pool and read data.
> > 
> > 
> > @rpadma2 @saurabhtandan, thoughts on this?
> 
> How often this problem happens under CI? Is it only in this test alone. I am not sure whether adding pool.destroy within the test should resolve the issue. If @saurabhtandan is ok with the change, we can land it... I will also check with @phender (Phil) to understand why it failed based on the existing logs. He can provide some input...

The idea of having pool destroy within the test is not to fix the issue but show the failure within the test, not the teardown. Regarding how often this problem appears, I don't think there is an easy way to query jenkins?

**Comment by phender** [comment]
> > > @omar-ocampo , any specific reason to add pool destroy as part of test run and not leave it for teardown?
> > > Also, I noticed the original issue in the ticket, created a list of pools and triggered rebuild and then on pool destroy there was a timeout.
> > > I understand with time the test might have evolved and there is no scenario in the test variants now where multiple pools are created as part of same test variant and then rebuild is triggered. (if I missed and there is one can you please point me to it)
> > > Do you think we should add that scenario and then test for pool destroy timeout?
> > 
> > 
> > Good question. The idea of moving the pool destroy within the test instead of the teardown is: if in the future we have the same/similar failure we can pinpoint the issue within the test and not the teardown. While the difference is subtle it may make a big difference in debugging. You are correct, if we speak in terms of functionality it would be the same (delete within test or teardown) I just favor the maintainability of the code.
> > Having say that, I'm open of letting the teardown to delete the pool, should we keep the pool delete within the test or moved it out?
> > Yes, only one pool is created. The rebuild testing is being done within pool/rebuild_tests.py, should be easier to add "--svn" as a test param and be able to use the logic of write data, stop server, rebuild pool and read data.
> 
> @rpadma2 @saurabhtandan, thoughts on this?

In my opinion it doesn't really matter if the pool destroy error/exception is in the test method or `tearDown()` - it's just less - and objectively cleaner - code to have `tearDown()` handle the pool destroy.  There is also the benefit of easier maintenance.  If we wanted to list some additional information when the pool destroy fails, we can easily add it in one place and affect multiple tests.  So my vote would be to leave the pool destroy in `tearDown()`.

**Comment by omar-ocampo** [comment]
> > > > @omar-ocampo , any specific reason to add pool destroy as part of test run and not leave it for teardown?
> > > > Also, I noticed the original issue in the ticket, created a list of pools and triggered rebuild and then on pool destroy there was a timeout.
> > > > I understand with time the test might have evolved and there is no scenario in the test variants now where multiple pools are created as part of same test variant and then rebuild is triggered. (if I missed and there is one can you please point me to it)
> > > > Do you think we should add that scenario and then test for pool destroy timeout?
> > > 
> > > 
> > > Good question. The idea of moving the pool destroy within the test instead of the teardown is: if in the future we have the same/similar failure we can pinpoint the issue within the test and not the teardown. While the difference is subtle it may make a big difference in debugging. You are correct, if we speak in terms of functionality it would be the same (delete within test or teardown) I just favor the maintainability of the code.
> > > Having say that, I'm open of letting the teardown to delete the pool, should we keep the pool delete within the test or moved it out?
> > > Yes, only one pool is created. The rebuild testing is being done within pool/rebuild_tests.py, should be easier to add "--svn" as a test param and be able to use the logic of write data, stop server, rebuild pool and read data.
> > 
> > 
> > @rpadma2 @saurabhtandan, thoughts on this?
> 
> In my opinion it doesn't really matter if the pool destroy error/exception is in the test method or `tearDown()` - it's just less - and objectively cleaner - code to have `tearDown()` handle the pool destroy. There is also the benefit of easier maintenance. If we wanted to list some additional information when the pool destroy fails, we can easily add it in one place and affect multiple tests. So my vote would be to leave the pool destroy in `tearDown()`.

Guess I'll remove the line and lets wait a day for the results, which should be the same

File: src/tests/ftest/pool/pool_svc.py [hunk 6]
```diff
@@ -130,6 +129,8 @@ def test_pool_svc(self):
                         "DaosServerManager.stop_ranks([{}])".format(
                             pool_leader))
 
+                self.pool.wait_for_rebuild(to_start=False, interval=1)
```
Comment (phender) [inline src/tests/ftest/pool/pool_svc.py:132]: To verify that rebuild has completed, this should be called twice - once to verify rebuild has indeed started (`to_start=True`) and a second time to ensure it has completed (`to_start=False`).  Without verifying that rebuild has started you may get a false positive when detecting if rebuild has completed.
```suggestion
                self.pool.wait_for_rebuild(to_start=True, interval=1)
                self.pool.wait_for_rebuild(to_start=False, interval=1)
```

You may also want to consider adding a `rebuild_timeout: <seconds>` yaml entry under the `pool:` section in this test's yaml file.  See https://github.com/daos-stack/daos/blob/1bc2be36b3b14cdc12bf349e4f52ac9174bf540e/src/tests/ftest/util/test_utils_pool.py#L484 for more details.

File: src/tests/ftest/pool/pool_svc.py [hunk 7]
```diff
@@ -148,6 +149,7 @@ def test_pool_svc(self):
                         "Error stopping a pool non-leader - "
                         "DaosServerManager.stop_ranks([{}])".format(non_leader))
 
+                self.pool.wait_for_rebuild(to_start=False, interval=1)
```
Comment (phender) [inline src/tests/ftest/pool/pool_svc.py:152]: Same comment as above.

**Review by phender — CHANGES_REQUESTED** [review]

**Review by shimizukko — APPROVED** [review]

**Review by phender — APPROVED** [review]

**Review by shimizukko — APPROVED** [review]

**Comment by omar-ocampo** [comment]
@sylviachanoiyee please merge, test results here: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5358/11/testReport/FTEST_pool/PoolSvc/

**Pull request merged**